{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "ab31bc6e",
   "metadata": {},
   "source": [
    "# Project 2D - Dylan Rhee"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "017356e1",
   "metadata": {},
   "source": [
    "### Scientific Question: What are the differences between the amino acid sequences in the top 3 proteins of whey protein (Beta-lactoglobulin, Alpha-lactalbumin, Glycomacropeptide) and the top 2 proteins in soybean (β-conglycinin, Glycinin)?\n",
    "Beta-lactoglobulin has been found to be one of the major components of whey protein and has been linked in many studies to increased cell proliferation. It has also been discovered that the conformation of beta-lactoglobulin plays a large role in determining its function. As a result, beta-lactoglobulin has been also considered in terms of its potential in enhancing human immunity (Tai et al., 2016). \n",
    "\n",
    "Another component of whey protein, alpha-lactalbumin, has been connected with many health benefits such as enhancing sleep, mood, immune responses, infant development, cancer treatment, and much more. The main component that has been studied that might give insight into these supposed health benefits is tryptophan especially in consideration of claimed benefits to mood and depression. In addition, there have been experiments done looking at connection of diets containing alpha-lactalbumin and how it increases tissue levels of gluthionine, which protects cells against oxidative stress (Layman et al., 2018).\n",
    "\n",
    "Last but not least for the whey proteins glycomacropeptide which has also boasted similar benefits such as prevention of some inflammatory diseases as well as cytoprotection against hydrogen peroxide. However, there are some unclear areas regarding glycomacropeptide's potential of anti-tumor properties and as a therapeutic. Glycomacropeptide shares a similar trait to beta-lactoglobulin where the structure plays a significant role in determining function (Córdova-Dávalos et al., 2019). \n",
    "\n",
    "Soybeans are largely comprised of two major storage proteins, β-conglycinin and glycinin, which comprise of 80-90% of the total protein in soybeans. Interestingly, soy peptides have been found to act as anti-hypertensives, anti-cancer agents, antioxidants and anti-inflammatories. It has been found that many soy peptides with hypolipidemic function, or the lowering of cholesterol, also possess anti-diabetic activity. Soy peptides have been found to have large associations with health benefits particularly the reduction of chronic diseases. To tests this numerous experiments have been done with diabetic rodents on a soy diet, which has been linked to improvements in their condition with fermented soy products showing even better results (Chatterjee et al., 2018).\n",
    "\n",
    "All of the data used within the scope of this experiment, the amino acid sequence of all five proteins, was pulled from the NCBI (National Center for Biotechnology Information: https://www.ncbi.nlm.nih.gov/). The specific database used was the protein database, which provides information such as taxonomy, source, the amino acid sequence, and more. For specific links as to where the data was found for each of the proteins they can be found under the \"Multiple Sequence Alignment Section\"."
   ]
  },
  {
   "cell_type": "markdown",
   "id": "b025acdd",
   "metadata": {},
   "source": [
    "### Hypothesis: If whey protein is considered to be more of a synthetic product as opposed to how soybean is seen as a more natural product, then there should little to no overlaps when comparing the amino acid sequences of the top 2-3 proteins found in each respective product. \n",
    "\n",
    "As stated previously all data was pulled from the NCBI protein database. From there, a small selection process was conducted for the protein such as taking into account things such as how likely the origin of the protein would be associated with human consumption. Afterwards, the fasta file would be downloaded and then read into the jupyter notebook as evident below in the multiple sequence alignment section.\n",
    "\n",
    "The first bioinformatics done was multiple sequence alignment. It should be noted that the packaged used required that all sequences be the same length and so gap sequences were needed to be added to the shorter sequences. Afterwards, multiple sequence alignment was conducted on all 5 protein samples. This would visually align all 5 protein sequences to determine if there was any common overlaps or not.\n",
    "\n",
    "The second bioinformatics method, pairiwise sequence alignment was conducted on every combination of two proteins regardless if both proteins were whey or both proteins were soy. This then yielded the score indicating the similarity of the two proteins being compared as well as visually showing the matches, mismatches, and gaps of the sequences.\n",
    "\n",
    "For the visualization of the data a heatmap was chosen to compare the scores for each of the protein combinations and then look for any patterns or relationships that can be interpreted. It should be noted no statistical test was done to determine if there is a statistically significant difference. Instead, the main focus was on the relative similarity betwen the different protein combinations to determine whether there were large overlaps or small overlaps in the amino acid sequence. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "d70eec5e",
   "metadata": {},
   "source": [
    "### Packages\n",
    "\n",
    "- Biopython: Biopython is an open-source collection of non-commercial Python tools for computational biology and bioinformatics. It is able to use a variety of file formats and deals with many types of data such as biological sequences and sequence annotations. Biopython has been designed to be functionally similar to other Bio projects and follows the conventions used by Python for ease of use. Biopython allows for accessing of online databases such as NCBI, BLAST, etc. Some modules of Biopython include SeqIO, AlignIO, pairiwise2, and more to do functions such as pairwise sequnce alignment and multiple sequence alignment. To learn more about Biopython see here: https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2682512/\n",
    "\n",
    "\n",
    "- Numpy: Numpy is the fundamental package for scientific computing in Python. Numpy adds support for large multi-dimensional arrays and matrices with a wide variety of mathematical functions to work with these arrays. One of the core functions of Numpy is the \"ndarray\" or n-dimensional array data structure as it is heavily used to perform many types of data manipulation or operations on a compiled set of data. The main benefit of Numpy is that it serves as a faster alternative compared to regular Python lists. In addition, Numpy is fast and versatile and is known as the defacto standards of array computing today. To learn more about Numpy see here: https://www.w3schools.com/python/numpy/numpy_random.asp\n",
    "\n",
    "\n",
    "- Matplotlib: Matplotlib serves as a helpful tool in data visualization and graphical plotting for Python. Matplotlib is helpful in visualizing a lot of data in a more digestible manner. Matplotlib includes various plots such as line, bar, scatter, histograms, pie charts, and more. In addition, Numpy serves as the numerical mathematics extension of Matplotlib and so they work together pretty much hand in hand. Matplotlib also works closely with pandas for data manipulation and analysis but is not a required dependency. To learn more about Matplotlib see here: https://www.activestate.com/resources/quick-reads/what-is-matplotlib-in-python-how-to-use-it-for-plotting/\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "aebbb83b",
   "metadata": {},
   "source": [
    "### Multiple Sequence Alignment \n",
    "Read in fasta files to be used for Multiple Sequence Alignment(MSA) but can use a number of file formats, which can be found here: https://biopython.org/wiki/AlignIO. A fasta file is a text file which beings with information about the sequence such as id, species, etc and is denoted by \">\" and the description following the \">\". The descripition is then followed by the actual sequence. MSA normally aims to compare 3 or more sequences by aligning the sequences together to be visually seen. From there, sequence homology can be inferred and other analyses can be performed.\n",
    "\n",
    "In the code below, fasta files are loaded and returned to confirm that fasta files are correct and loaded properly. Then the fasta files are compiled into a larger fasta file while first making sure all sequences are the same length by adding gap sequences. The compiled fasta file will then be used to conduct multiple sequence alignment. "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "dbcaf332",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Import Biopython so that module such as AlignIO, SeqIO, and pairiwise2 can be imported later on\n",
    "import Bio\n",
    "\n",
    "#import ALignIO for multiple sequence alignment\n",
    "#Source: https://biopython.org/wiki/AlignIO\n",
    "from Bio import AlignIO "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "86f775f7",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Alignment with 1 rows and 180 columns\n",
      "MKCLLLALGLALACGAQAIIVTQTMKGLDIQKVAGTWYSLAMAA...CHV CAY39357.1\n",
      "Alignment with 1 rows and 142 columns\n",
      "MRFFVPLFLVGILFPAILAKQFTKCELSQLLKDIDGYGGIALPE...EKL BAC06860.1\n",
      "Alignment with 1 rows and 190 columns\n",
      "MMKSFFLVVTILALTLPFLGAQEQNQEQPIRCEKDERFFSDKIA...TAV sp|P02668.1|CASK_BOVIN\n",
      "Alignment with 1 rows and 439 columns\n",
      "MMRVRFPLLVLLGTVFLASVCVSLKVREDENNPFYFRSSNSFQT...ALY AAB23463.1\n",
      "Alignment with 1 rows and 516 columns\n",
      "MGKPFFTLSLSSLCLLLLSSACFAITSSKFNECQLNNLNALEPD...VNP AAA33964.1\n"
     ]
    }
   ],
   "source": [
    "#Beta-lactoglobulin = Whey1\n",
    "#Alpha-lactalbumin = Whey2\n",
    "#Glycomacropeptide = Whey3\n",
    "#β-conglycinin = Soy1\n",
    "#Glycinin = Soy2 \n",
    "#variable = AlignIO.read(\"filename\", \"filetype\")\n",
    "\n",
    "#The following code shows how a file containing single alignment is loaded and returns the amino acid sequence of the protein followed by protein id\n",
    "#The print functions serve as code checks to make sure the files have the correct information and are returning a sequence and sequence id\n",
    "\n",
    "#Fasta File for Beta-lactoglobulin: https://www.ncbi.nlm.nih.gov/protein/CAY39357.1 \n",
    "#Whey1 is a global variable but won't be used for remainder portion of code\n",
    "Whey1 = AlignIO.read(\"beta_lactoglobulin\", \"fasta\")\n",
    "print(Whey1)\n",
    "\n",
    "#Fasta File for Alpha-lactalbumin: https://www.ncbi.nlm.nih.gov/protein/BAC06860.1\n",
    "#Whey2 is a global variable but won't be used for remainder portion of code\n",
    "Whey2 = AlignIO.read(\"alpha_lactalbumin\", \"fasta\")\n",
    "print(Whey2)\n",
    "\n",
    "#Fasta File for Glycomacropeptide: https://www.ncbi.nlm.nih.gov/protein/P02668\n",
    "#Whey3 is a global variable but won't be used for remainder portion of code\n",
    "Whey3 = AlignIO.read(\"glycomacropeptide\", \"fasta\")\n",
    "print(Whey3)\n",
    "\n",
    "#Fasta File for β-conglycinin: https://www.ncbi.nlm.nih.gov/protein/AAB23463.1\n",
    "#Soy1 is a global variable but won't be used for remainder portion of code\n",
    "Soy1 = AlignIO.read(\"beta_conglycinin\", \"fasta\")\n",
    "print(Soy1)\n",
    "\n",
    "#Fasta File for Glycinin: https://www.ncbi.nlm.nih.gov/protein/AAA33964.1\n",
    "#Soy2 is a global variable but won't be used for remainder portion of code\n",
    "Soy2 = AlignIO.read(\"glycinin\", \"fasta\")\n",
    "print(Soy2)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "b725bf87",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "['CAY39357.1', 'BAC06860.1', 'sp|P02668.1|CASK_BOVIN', 'AAB23463.1', 'AAA33964.1']\n",
      "['MKCLLLALGLALACGAQAIIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEELKPTPEGDLEILLQKWENGECAQKKIIAEKTKIPAVFKIDALNENKVLVLDTDYKKYLLFCMENSAEPEQSLACQCLVRTPEVDDEALEKFDKALKALPMHIRLSFNPTQLEEQCHV', 'MRFFVPLFLVGILFPAILAKQFTKCELSQLLKDIDGYGGIALPELICTMFHTSGYDTQAIVENNESTEYGLFQISNKLWCKSSQVPQSRNICDISCDKFLDDDITDDIMCAKKILDIKGIDYWLAHKALCTEKLEQWLCEKL', 'MMKSFFLVVTILALTLPFLGAQEQNQEQPIRCEKDERFFSDKIAKYIPIQYVLSRYPSYGLNYYQQKPVALINNQFLPYPYYAKPAAVRSPAQILQWQVLSNTVPAKSCQAQPTTMARHPHPHLSFMAIPPKKNQDKTEIPTINTIASGEPTSTPTTEAVESTVATLEDSPEVIESPPEINTVQVTSTAV', 'MMRVRFPLLVLLGTVFLASVCVSLKVREDENNPFYFRSSNSFQTLFENQNVRIRLLQRFNKRSPQLENLRDYRIVQFQSKPNTILLPHHADADFLLFVLSGRAILTLVNNDDRDSYNLHPGDAQRIPAGTTYYLVNPHDHQNLKIIKLAIPVNKPGRYDDFFLSSTQAQQSYLQGFSHNILETSFHSEFEEINRVLFGEEEEQRQQEGVIVELSKEQIRQLSRRAKSSSRKTISSEDEPFNLRSRNPIYSNNFGKFFEITPEKNPQLRDLDIFLSSVDINEGALLLPHFNSKAIVILVINEGDANIELVGIKEQQQKQKQEEEPLEVQRYRAELSEDDVFVIPAAYPFVVNATSNLNFLAFGINAENNQRNFLAGEKDNVVRQIERQVQELAFPGSAQDVERLLKKQRESYFVDAQPQQKEEGSKGRKGPFPSILGALY', 'MGKPFFTLSLSSLCLLLLSSACFAITSSKFNECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRTLNRNGSHLPSYLPYPQMIIVVQGKGAIGFAFPGCPETFEKPQQQSSRRGSRSQQQLQDSHQKIRHFNEGDVLVIPLGVPYWTYNTGDEPVVAISPLDTSNFNNQLDQNPRVFYLAGNPDIEHPETMQQQQQQKSHGGRKQGQHRQQEEEGGSVLSGFSKHFLAQSFNTNEDTAEKLRSPDDERKQIVTVEGGLSVISPKWQEQEDEDEDEDEEYGRTPSYPPRRPSHGKHEDDEDEDEEEDQPRPDHPPQRPSRPEQQEPRGRGCQTRNGVEENICTMKLHENIARPSRADFYNPKAGRISTLNSLTLPALRQFGLSAQYVVLYRNGIYSPDWNLNANSVTMTRGKGRVRVVNCQGNAVFDGELRRGQLLVVPQNPAVAEQGGEQGLEYVVFKTHHNAVSSYIKDVFRVIPSEVLSNSYNLGQSQVRQLKYQGNSGPLVNP']\n"
     ]
    }
   ],
   "source": [
    "#Goal: Make a new fasta file that compiles the 5 fasta files for each indiivudal protein\n",
    "#Essentially combining the fasta files \n",
    "#First have to put names and sequences into two different lists\n",
    "#Source: https://www.biostars.org/p/18129/\n",
    "\n",
    "#Import SeqIO to help with appending header and sequences from the fasta files \n",
    "from Bio import SeqIO\n",
    "\n",
    "#Defining empty lists for the headers (headerList) and sequences (seqList)\n",
    "headerList = []\n",
    "seqList = []\n",
    "\n",
    "#Appending header and sequence from each individual protein fasta file to the empty lists defined previously\n",
    "#Whey1File is a global variable that contains header and sequences that will then be apppended to headerList and seqList respectively\n",
    "Whey1File = open('beta_lactoglobulin', 'r')\n",
    "for record in SeqIO.parse(Whey1File,'fasta'):\n",
    "   headerList.append(record.id)\n",
    "   seqList.append(str(record.seq))\n",
    "    \n",
    "#Whey2File is a global variable that contains header and sequences that will then be apppended to headerList and seqList respectively \n",
    "Whey2File = open('alpha_lactalbumin', 'r')\n",
    "for record in SeqIO.parse(Whey2File,'fasta'):\n",
    "   headerList.append(record.id)\n",
    "   seqList.append(str(record.seq))\n",
    "\n",
    "#Whey3File is a global variable that contains header and sequences that will then be apppended to headerList and seqList respectively    \n",
    "Whey3File = open('glycomacropeptide', 'r')\n",
    "for record in SeqIO.parse(Whey3File,'fasta'):\n",
    "   headerList.append(record.id)\n",
    "   seqList.append(str(record.seq))\n",
    "\n",
    "#Soy1File is a global variable that contains header and sequences that will then be apppended to headerList and seqList respectively\n",
    "Soy1File = open('beta_conglycinin', 'r')\n",
    "for record in SeqIO.parse(Soy1File,'fasta'):\n",
    "   headerList.append(record.id)\n",
    "   seqList.append(str(record.seq))\n",
    "    \n",
    "#Soy2File is a global variable that contains header and sequences that will then be apppended to headerList and seqList respectively   \n",
    "Soy2File = open('glycinin', 'r')\n",
    "for record in SeqIO.parse(Soy2File,'fasta'):\n",
    "   headerList.append(record.id)\n",
    "   seqList.append(str(record.seq))\n",
    "\n",
    "#Code check checking to see if the lists were made and are correct\n",
    "print(headerList)\n",
    "print(seqList)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "b7615b93",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "516\n",
      "['MKCLLLALGLALACGAQAIIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEELKPTPEGDLEILLQKWENGECAQKKIIAEKTKIPAVFKIDALNENKVLVLDTDYKKYLLFCMENSAEPEQSLACQCLVRTPEVDDEALEKFDKALKALPMHIRLSFNPTQLEEQCHV------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------', 'MRFFVPLFLVGILFPAILAKQFTKCELSQLLKDIDGYGGIALPELICTMFHTSGYDTQAIVENNESTEYGLFQISNKLWCKSSQVPQSRNICDISCDKFLDDDITDDIMCAKKILDIKGIDYWLAHKALCTEKLEQWLCEKL--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------', 'MMKSFFLVVTILALTLPFLGAQEQNQEQPIRCEKDERFFSDKIAKYIPIQYVLSRYPSYGLNYYQQKPVALINNQFLPYPYYAKPAAVRSPAQILQWQVLSNTVPAKSCQAQPTTMARHPHPHLSFMAIPPKKNQDKTEIPTINTIASGEPTSTPTTEAVESTVATLEDSPEVIESPPEINTVQVTSTAV--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------', 'MMRVRFPLLVLLGTVFLASVCVSLKVREDENNPFYFRSSNSFQTLFENQNVRIRLLQRFNKRSPQLENLRDYRIVQFQSKPNTILLPHHADADFLLFVLSGRAILTLVNNDDRDSYNLHPGDAQRIPAGTTYYLVNPHDHQNLKIIKLAIPVNKPGRYDDFFLSSTQAQQSYLQGFSHNILETSFHSEFEEINRVLFGEEEEQRQQEGVIVELSKEQIRQLSRRAKSSSRKTISSEDEPFNLRSRNPIYSNNFGKFFEITPEKNPQLRDLDIFLSSVDINEGALLLPHFNSKAIVILVINEGDANIELVGIKEQQQKQKQEEEPLEVQRYRAELSEDDVFVIPAAYPFVVNATSNLNFLAFGINAENNQRNFLAGEKDNVVRQIERQVQELAFPGSAQDVERLLKKQRESYFVDAQPQQKEEGSKGRKGPFPSILGALY-----------------------------------------------------------------------------', 'MGKPFFTLSLSSLCLLLLSSACFAITSSKFNECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRTLNRNGSHLPSYLPYPQMIIVVQGKGAIGFAFPGCPETFEKPQQQSSRRGSRSQQQLQDSHQKIRHFNEGDVLVIPLGVPYWTYNTGDEPVVAISPLDTSNFNNQLDQNPRVFYLAGNPDIEHPETMQQQQQQKSHGGRKQGQHRQQEEEGGSVLSGFSKHFLAQSFNTNEDTAEKLRSPDDERKQIVTVEGGLSVISPKWQEQEDEDEDEDEEYGRTPSYPPRRPSHGKHEDDEDEDEEEDQPRPDHPPQRPSRPEQQEPRGRGCQTRNGVEENICTMKLHENIARPSRADFYNPKAGRISTLNSLTLPALRQFGLSAQYVVLYRNGIYSPDWNLNANSVTMTRGKGRVRVVNCQGNAVFDGELRRGQLLVVPQNPAVAEQGGEQGLEYVVFKTHHNAVSSYIKDVFRVIPSEVLSNSYNLGQSQVRQLKYQGNSGPLVNP']\n"
     ]
    }
   ],
   "source": [
    "#Adding gap characters so all proteins have same length, necessary for later when doing multiple sequence alignment\n",
    "#Source: https://stackoverflow.com/questions/31152011/multiple-sequence-alignment-with-unequal-string-length\n",
    "\n",
    "#Figuring out which protein has the longest amino acid length by looking at max length of all sequences in seqList\n",
    "#longest is a global variable\n",
    "longest = max(len(s) for s in seqList)\n",
    "print(longest)\n",
    "\n",
    "#Adding gaps to other proteins to match the protein with the longest amino acid length in global variable \"longest\"\n",
    "gaps = [s.ljust(longest, '-') for s in seqList]\n",
    "#Code check to see that gap sequences were added \n",
    "print(gaps)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "c2861f3f",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create the actual file using the two lists we made for the header and sequence\n",
    "#Source: https://www.biostars.org/p/142519/\n",
    "\n",
    "#Note: gaps is the updated list containing the sequences and so seqList is not used\n",
    "#Creating a new file where we will be adding the header and sequence for each of the proteins to get our compiled fasta file\n",
    "ofile = open(\"Comparison1.txt\", \"w\")\n",
    "for i in range(len(gaps)):\n",
    "    ofile.write(\">\" + headerList[i] + \"\\n\" +gaps[i] + \"\\n\") \n",
    "ofile.close()\n",
    "\n",
    "#Create a file where gap characters were not added, this will later be used for pairwise sequence alignment \n",
    "ofile = open(\"Comparison2.txt\", \"w\")\n",
    "for i in range(len(seqList)):\n",
    "    ofile.write(\">\" + headerList[i] + \"\\n\" +seqList[i] + \"\\n\") \n",
    "ofile.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "3ddfc2af",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "MKCLLLALGLALACGAQAIIVTQTMKGLDIQKVAGTWYSLAMAASDISLLDAQSAPLRVYVEELKPTPEGDLEILLQKWENGECAQKKIIAEKTKIPAVFKIDALNENKVLVLDTDYKKYLLFCMENSAEPEQSLACQCLVRTPEVDDEALEKFDKALKALPMHIRLSFNPTQLEEQCHV------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "MRFFVPLFLVGILFPAILAKQFTKCELSQLLKDIDGYGGIALPELICTMFHTSGYDTQAIVENNESTEYGLFQISNKLWCKSSQVPQSRNICDISCDKFLDDDITDDIMCAKKILDIKGIDYWLAHKALCTEKLEQWLCEKL--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "MMKSFFLVVTILALTLPFLGAQEQNQEQPIRCEKDERFFSDKIAKYIPIQYVLSRYPSYGLNYYQQKPVALINNQFLPYPYYAKPAAVRSPAQILQWQVLSNTVPAKSCQAQPTTMARHPHPHLSFMAIPPKKNQDKTEIPTINTIASGEPTSTPTTEAVESTVATLEDSPEVIESPPEINTVQVTSTAV--------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------\n",
      "MMRVRFPLLVLLGTVFLASVCVSLKVREDENNPFYFRSSNSFQTLFENQNVRIRLLQRFNKRSPQLENLRDYRIVQFQSKPNTILLPHHADADFLLFVLSGRAILTLVNNDDRDSYNLHPGDAQRIPAGTTYYLVNPHDHQNLKIIKLAIPVNKPGRYDDFFLSSTQAQQSYLQGFSHNILETSFHSEFEEINRVLFGEEEEQRQQEGVIVELSKEQIRQLSRRAKSSSRKTISSEDEPFNLRSRNPIYSNNFGKFFEITPEKNPQLRDLDIFLSSVDINEGALLLPHFNSKAIVILVINEGDANIELVGIKEQQQKQKQEEEPLEVQRYRAELSEDDVFVIPAAYPFVVNATSNLNFLAFGINAENNQRNFLAGEKDNVVRQIERQVQELAFPGSAQDVERLLKKQRESYFVDAQPQQKEEGSKGRKGPFPSILGALY-----------------------------------------------------------------------------\n",
      "MGKPFFTLSLSSLCLLLLSSACFAITSSKFNECQLNNLNALEPDHRVESEGGLIETWNSQHPELQCAGVTVSKRTLNRNGSHLPSYLPYPQMIIVVQGKGAIGFAFPGCPETFEKPQQQSSRRGSRSQQQLQDSHQKIRHFNEGDVLVIPLGVPYWTYNTGDEPVVAISPLDTSNFNNQLDQNPRVFYLAGNPDIEHPETMQQQQQQKSHGGRKQGQHRQQEEEGGSVLSGFSKHFLAQSFNTNEDTAEKLRSPDDERKQIVTVEGGLSVISPKWQEQEDEDEDEDEEYGRTPSYPPRRPSHGKHEDDEDEDEEEDQPRPDHPPQRPSRPEQQEPRGRGCQTRNGVEENICTMKLHENIARPSRADFYNPKAGRISTLNSLTLPALRQFGLSAQYVVLYRNGIYSPDWNLNANSVTMTRGKGRVRVVNCQGNAVFDGELRRGQLLVVPQNPAVAEQGGEQGLEYVVFKTHHNAVSSYIKDVFRVIPSEVLSNSYNLGQSQVRQLKYQGNSGPLVNP\n"
     ]
    }
   ],
   "source": [
    "#Multiple Sequence Alignment looking to compare all proteins: Whey1, Whey2, Whey3, Soy1, and Soy2\n",
    "\n",
    "#Global variable alignment that contains the multiple sequence alignment for \"Comparison1.txt\"\n",
    "alignment = AlignIO.read(open(\"Comparison1.txt\"), \"fasta\")\n",
    "#Code check \n",
    "#print(alignment)\n",
    "\n",
    "#Code check to make sure alignment legnth is correct\n",
    "#print(\"Alignment length %i\" % alignment.get_alignment_length())\n",
    "\n",
    "#Visualize the alignments of the proteins and associated names by using a for loop to print every sequence and sequence id if needed\n",
    "for record in alignment:\n",
    "    #First print function will give the sequence followed by id of the sequence. The second print function will only show the sequences for the five proteins\n",
    "    #print(record.seq + \" \" + record.id)\n",
    "    print(record.seq)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "566fea04",
   "metadata": {},
   "source": [
    "### Pairwise Sequence Alignment \n",
    "Similar to Multiple Sequence Alignment, in the Pairwise Sequence Alignment(PSA) shown below a fasta file was loaded in. The fasta file is a regular text file that contains a header denoted by \">\" followed by information regarding the identity and more. Following the header is the sequence of the protein that will later be used in PSA. PSA aims to look at the matches, mismatches, and gaps between two sequences where each is assigned a respective value that will then be used to calculate the score. The score is indicative of how homologous the two sequences are to one another with a higher score indicating increased homology. \n",
    "\n",
    "In the code below there are two sections where PSA is done. The first section looks to individually examine PSA between all possible protein combinations. The first section is redundant but really helps when wanting to visualize the possible matches/mismatches/gaps between two proteins. The second section creates an empty numpy matrix that will be filled with scores after PSA is done on between all protein combinations using a for loop. The second section will later then be used for the heatmap."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "35b7eff9",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Goal is to do pairwise sequence alignment between every possible protein pair\n",
    "#Note: This way of doing pairwise is a little redundant in terms of the code and so is only shown to visualize the matches/mismatches/gaps\n",
    "#Sources: https://towardsdatascience.com/pairwise-sequence-alignment-using-biopython-d1a9d0ba861f\n",
    "\n",
    "#import pairwise2 which will be used to do pairwise sequence alignment\n",
    "from Bio import pairwise2\n",
    "\n",
    "#import format_alignment to help with visualizing the matches/mismatches/gaps\n",
    "from Bio.pairwise2 import format_alignment\n",
    "\n",
    "#Note: have previous list seqList with all sequences and can check with print function also don't want to use gaps because pairwise sequence alignment will matches the gaps together\n",
    "#Beta-lactoglobulin = seqList[0]\n",
    "#Alpha-lactalbumin = seqList[1]\n",
    "#Glycomacropeptide = seqList[2]\n",
    "#β-conglycinin = seqList[3]\n",
    "#Glycinin = seqList[4]\n",
    "#Code check/reminder\n",
    "#print(seqList)\n",
    "\n",
    "#Note this is not the most efficient method of doing the code but helps with visualizing the various matches/mismatches/gaps if you wanna focus on an individual comparison for the proteins. For more efficient code refer to the section where pairwise is done with the for loop below.\n",
    "#Function that does pairwise sequence alignment on two sequences\n",
    "#pairwise_1(global variable) = Beta-lactoglobulin and Alpha-lactalbumin \n",
    "pairwise_1 = pairwise2.align.globalxx(seqList[0], seqList[1])\n",
    "#Code check\n",
    "#for a in pairwise_1:\n",
    "    #print(format_alignment(*a))\n",
    "    \n",
    "#pairwise_2(global variable) = Beta-lactoglobulin and Glycomacropeptide \n",
    "pairwise_2 = pairwise2.align.globalxx(seqList[0], seqList[2])\n",
    "#Code check\n",
    "#for a in pairwise_2:\n",
    "    #print(format_alignment(*a))\n",
    "    \n",
    "#pairwise_3(global variable) = Beta-lactoglobulin and β-conglycinin \n",
    "pairwise_3 = pairwise2.align.globalxx(seqList[0], seqList[3])\n",
    "#Code check\n",
    "#for a in pairwise_3:\n",
    "    #print(format_alignment(*a))\n",
    "    \n",
    "#pairwise_4(global variable) = Beta-lactoglobulin and Glycinin\n",
    "pairwise_4 = pairwise2.align.globalxx(seqList[0], seqList[4])\n",
    "#Code check\n",
    "#for a in pairwise_4:\n",
    "    #print(format_alignment(*a))\n",
    "    \n",
    "#pairwise_5(global variable) = Alpha-lactalbumin and Glycomacropeptide\n",
    "pairwise_5 = pairwise2.align.globalxx(seqList[1], seqList[2])\n",
    "#Code check\n",
    "#for a in pairwise_5:\n",
    "    #print(format_alignment(*a))\n",
    "\n",
    "#pairwise_6(global variable) = Alpha-lactalbumin and β-conglycinin \n",
    "pairwise_6 = pairwise2.align.globalxx(seqList[1], seqList[3])\n",
    "#Code check\n",
    "#for a in pairwise_6:\n",
    "    #print(format_alignment(*a))\n",
    "\n",
    "#pairwise_7(global variable) = Alpha-lactalbumin and Glycinin\n",
    "pairwise_7 = pairwise2.align.globalxx(seqList[1], seqList[4])\n",
    "#Code check\n",
    "#for a in pairwise_7:\n",
    "    #print(format_alignment(*a))\n",
    "\n",
    "#pairwise_8(global variable) = Glycomacropeptide and β-conglycinin\n",
    "pairwise_8 = pairwise2.align.globalxx(seqList[2], seqList[3])\n",
    "#Code check\n",
    "#for a in pairwise_8:\n",
    "    #print(format_alignment(*a))\n",
    "\n",
    "#pairwise_9(global variable) = Glycomacropeptide and Glycinin\n",
    "pairwise_9 = pairwise2.align.globalxx(seqList[2], seqList[4])\n",
    "#Code check\n",
    "#for a in pairwise_9:\n",
    "    #print(format_alignment(*a))\n",
    "\n",
    "#pairiwse_10(global variable) = β-conglycinin and Glycinin\n",
    "pairwise_10 = pairwise2.align.globalxx(seqList[3], seqList[4])\n",
    "#Code check\n",
    "#for a in pairwise_10:\n",
    "    #print(format_alignment(*a))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "2782b3d8",
   "metadata": {},
   "outputs": [],
   "source": [
    "#Pairwise Sequence Alignment on Comparison2.txt which is the fasta file cotaining the sequences of all five proteins without gaps\n",
    "#Source: https://canvas.ucsd.edu/courses/29418/files/5721198?module_item_id=1126617&fd_cookie_set=1\n",
    "\n",
    "#Import numpy to be used when creating the numpy matrix\n",
    "import numpy as np\n",
    "\n",
    "#Put all data from Comparison2.txt into a list such as header, amino acid sequence, and some other information\n",
    "#Global variable PSA\n",
    "PSA = list(SeqIO.parse(\"Comparison2.txt\", \"fasta\"))\n",
    "\n",
    "#Code check to see that all the information was put into a list \n",
    "#print(PSA)\n",
    "\n",
    "#Get length of the list\n",
    "#Global variable size\n",
    "size = len(PSA)\n",
    "\n",
    "#Define variable i and j for the for loop\n",
    "i = 0\n",
    "j = 0\n",
    "\n",
    "#Create an empty numpy matrix wth dimensions that matches the length of PSA\n",
    "#Global variable score\n",
    "score = np.empty(shape=(size,size))\n",
    "\n",
    "#Nested for loop to do pairwise sequence alignment for every combination of the proteins  \n",
    "#Defining the function \"pairwise\" with parameters of PSA_array which is the empty array where the scores will be put in after doing pairwise sequence alignment and size to make sure that the array created is the correct dimensions \n",
    "def pairwise(PSA_array, size):\n",
    "    #this part of for loop helps run through every protein to be compared with all other proteins including itself\n",
    "    for i in range(size):\n",
    "        for j in range(size):\n",
    "        #Takes the sequence portion of the list and puts them into a local variable to then do pairwise later\n",
    "            X = PSA[i].seq\n",
    "            Y = PSA[j].seq\n",
    "\n",
    "        # Creates an array of all the scores after all pairwise between every protein has been done and returns only the score (numerical value)\n",
    "        # No parameters. Matches have a score of 1 while mismatch and gaps are 0.\n",
    "            PSA_array[i,j] = int(pairwise2.align.globalxx(X, Y, score_only=True))\n",
    "#Returns \"PSA_array\" which is the updated matrix including all scores after doing pairwise sequence alignment\n",
    "    return PSA_array\n",
    "#Global variable heatmapscore which fills an empty array with scores from pairwise sequence alignemtn on proteins from \"PSA\" with appropriate dimensions\n",
    "heatmapscore = pairwise(score, size)\n",
    "#Code check to see that the matrix of scores was created correctly\n",
    "#print(heatmapscore)\n"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c7067678",
   "metadata": {},
   "source": [
    "### Heat Plot\n",
    "A heat plot is a data visualization method that uses a color gradient to depict the magnitude of a particular value in two dimensions. In this case the paritcular value is the pairiwise sequence alignment scores between the two proteins being compared. \n",
    "\n",
    "In the code below the heat plot was made with the pairwise sequence alignment scores from above. Matplotlib was imported to help with setting up the heat plot and doing the actually visualization. In addition, axes were added with labels as well as labels for xticks and yticks. A colorbar was added to show what values correspond to the gradient of colors. Finally a title was added to help with distinguishing what the heat plot is depicting. However, it was necessary to normalize the scores considering each protein was different length and so the pairwise sequence alignment scores were normalized for each column."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "8d2e2349",
   "metadata": {
    "scrolled": true
   },
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "[[180.  58.  68.  96. 101.]\n",
      " [ 58. 142.  53.  88.  84.]\n",
      " [ 68.  53. 190.  97. 102.]\n",
      " [ 96.  88.  97. 439. 176.]\n",
      " [101.  84. 102. 176. 516.]]\n"
     ]
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAVIAAAEWCAYAAAA0HB+VAAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuNCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy8QVMy6AAAACXBIWXMAAAsTAAALEwEAmpwYAAAruklEQVR4nO3debwcVZ338c/3XsISAwQNIhI2ISxhCxDBRAYQRAFhIo4z4IIQZCLOBGXkccSBUWZcHlFRRMFMhAAqmyhowCioEMKwSAiEJUgwTxASgkjCHpaQ8Hv+OKeh0vTt2zfd93Z18n3n1a90LX3qV7e7f33qVNU5igjMzGzVdbU7ADOzTudEambWJCdSM7MmOZGamTXJidTMrElOpGZmTVojEqmk30g6pgXl/Iek81oRk/VO0rGS/rcw/bykd7QzpnaTtJWkkLRWu2Ox13VUIpX0F0kv5i/U45IukDSkt9dFxCERcVGz24+Ir0fE8c2WUyRpqKQpkv4q6TlJD0r6Qiu3UXaSTs/JYa9660XEkIiYP1BxtYKkCyV9tc7y/5F0bmF6kKSlPcx7V3/H22qSdpJ0naSnJD0taZakQ9sdV6t1VCLNDo+IIcAewDuB05oprAS/7N8FhgA7AhsCfw/8v7ZGNIAkCTgaeBJo+qihA80A9itMjwYeAfatmgcwa6CCaqGrgd8BmwBvBT4DPNvWiPpDRHTMA/gL8N7C9LeAa4CN8v9PAE/l58ML600Hjs/PjwVuJiWwJ4GvAg8De+blHwcCGJmnjwd+mZ+fDvw0P18X+CmwBHgamAlskpdtCJwPPAY8mrfR3cM+3Qd8sM4+70D6ID4JzAX+qbDsLcBU0gfzduArwP/mZVvl/Vir1t8hTx8H/Cn/za4FtiwsC+AE4M95+TmACsv/Ob/2OeB+YI88/+3AL/J78RDwmV7e032BF/PffQmwdmHZsZX9KcS0bWHfr877PjP/javXrRl/1WfgaWA+MDbPXwD8DTimUNY6wLdJCe5xYBKwXl62P7AQODm/7jFgfF42AXgFWAY8D1xdY/+HA68Cw/L0vwNfzn+74rzfV72vx+R4FgOnFsrrAk4h/RgvAX4GvDkv+zVwYtX276HG5w/4LTCxat7dwIcA5b/d34Bnchk71yhjWI51aJ33/5+BeaTP91Tg7Xn+OcCZVeteDZzU7jxUcz/aHUCfgi0kUmBzYA4pebwF+AdgMLA+cAU5+eV1p7NyIl0OnAisBawH/Bg4OS+fnD+En87TPwb+LT8/ndcT6afyGzsY6Ab2BDbIy34J/A/wJtKv8O3Ap3rYp/PyfowHRlQtexPpiz0+x7pH/uLslJdflr8obwJ2JiXthhIp8MH8Ad4xl30acEth3SD9IA0FtiAlxoPzsn/M23pn/lJtC2xJ+hLPAr4ErA28g5Sk3l/nPT0/78Mg0hf/Q4Vlx9JzIr0sPwYDI/PfqXrdnuKvfAbG5/fuq6SkdA4pab6P9AMxJK9/FulL/mbS5+tq4P/mZfvnsv4778OhwAvARnn5hcBXe/lcPwQckZ9fAxwAXFw170tV7+uPSJ/d3YCXgR3z8pOA20gJeh3S5/DSvOyfgD8WtrsbVT9ehWWfAG4uTI8k/eisA7w/v89D8/u/I7BpjTJE+iG7hvR526Rq+QGkz/MeudzvAzPysr2ARUBXnh6W/66b1Ptbti03tTuAPgWbEunz+Q19GDiXXDOoWm8U8FRhejorJ9JHqtb/JDA1P/8TqRZ6WZ5+mNdrW6fzeiI9DrgF2LWqrE3yB3u9wryPADf0sE/rAf+RP5ivkJLbIXnZkcBNVev/D6nG0p3X36Gw7Os0nkh/A3yysKwrf1C3zNMB7FNY/jPglPz8WuCzNfZl7xp/2y8CF/Sw74NJNcoPFvbtV4Xlx1IjkRb2ffvCslo10p7iPxb4c2HZLnn9TQrzluTPkYClwDaFZWOAh/Lz/Uk16uLf+W/Au/LzC+k9kV5IquF15dcOJtWmK/OeAvarel+LR1y3A0cVPr8HFpZtmv9Wa5GS1ZPkH2xSLfvcHmJaP+935fPwNWBKfn4A8CDwLnKiq7Nvw4EfkConr5KaMirbPx/4ZmHdITnWrQr7clB+PhGYtip5YyAendhG+sGIGBoRW0bEv0TEi5IG50b7hyU9S3qzhkrq7qGMBVXTNwJ/J+ltpC/p5cC7JW1FOkyfXaOMn5ASymWSFkn6pqRBpJrZIOCx3Lj+NClBvLVWIBHxYqSTWHuSatY/A66Q9OZc1t6VcnJZHwPeBmxM+nIU9+XhHva3li2B7xXKfZKUNDYrrPPXwvMXSB90SEcDtdpxtwTeXhXvf5B+XGo5glSbm5anLwYOkbRxL7HX2vfq97Re/JAO0SteBIiI6nlD8rYGA7MK+/TbPL9iSUQsr7Ot3swgNXHsAsyPiBeA/y3MWw/4Y4P7tiVwVSHWPwErSD8SL5M+Xx+X1EX6gf9JrYAi4jlSU8BRedZRpPeHiLielBzPAR6XNFnSBj2UszAiJkbENjm2paSjPEjNQA8X1n2e9ANW+QxeRGryIf9fM9Yy6MREWsvJwPbA3hGxAa831KuH9WOliYh5pA/jZ0iHFs+RPqgTSLWcV99QQMQrEfFfETGS1L52GOlwaAGpRjosJ/yhEbFBROzU205ExLOkWuWbgK1zWTcWyhka6cz1p0mHqstJSa1ii8Lzpfn/wYV5bys8X0BqbiiWvV5E3NJbnPm12/Qw/6GqMtePiJ7O0h5DSgCPSPorqUlmEOkLXk9l34cX5m3ew7rNWkxKqjsV9mnDSCc8GxG9r8IM0mH2B4Cb8rw5pH36ADAzIl5qcHsLSEc0xfdg3Yh4NC+/iPRjfCDwQkTcWqesS4GPSBpDSuY3vLZTEWfnH/+dgO2Az/cWWEQsICXfnfOsRaTkCoCkN5EqE5VYfwqMk7Qbqfngl71to11Wl0S6PunD/nSuyX15Fcq4kXT4cGOenl41vRJJ75G0S671Pks6JFkREY8B1wFnStpAUpekbSTt10M5/ynpnZLWlrQu8FlS08VcUtvSdpKOzpfADMrr7hgRK4ArgdNzjXwkhbPeEfEE6QP5cUndko5j5eQ3CfiipJ1yHBtK+scG/1bnAf9H0p5KtpW0JekQ81lJX5C0Xt7uzpLeWWO/NyN9mQ8jHUKPIiWTM+jl7H2Nfd+B9CPWcvlH9EfAdyW9tRK7pPc3WMTjpLbietuYl9f7LDmRRjqe/WOeN6MPIU8CvpbfDyRtLGlcYVu3kg6xz6T3Gt40UqL7b+DySoUifwb3zkdgS4GXSLXelUjaSNJ/5c9Hl6RhpCax2/IqlwDjJY2StA6pEvHHiPhLjnUh6UTiT4BfRMSLffg7DKjVJZGeRfrFXEx6k367CmXcSErIM3qYrvY24OekJPqnvP5P87JPkE623E9q3/o5qa2qlgAuyLEvAg4CPhARz+ea8ftIh1WLSLXkM0htXZAS/ZA8/8JcTtE/k2oKS0g1h9dqmxFxVS7rstwcch9wSA8xrhxwxBWkNrNLSCdlfkk6M7wCOJyUFB/K+3QeqXmk2tHA7Ii4LiL+WnkAZwO7Stq5xmuKJuZy/0r6ol1KOhLoD18gtV3flv9WvycdATXifGBkPtT+ZZ31ZpCaC24uzLuJ1CTUl0T6PdKJseskPUf6Puxdtc6PSU0GP6WO3BRwJfBe0ntdsQHpx+Up0qH5ElJ7a7VlpDbd35O+J/eR3qNjc/l/AP6TdJXHY6Qf+qOqyrgox1raw3p4/XIQWw1IOpZ0Mmmfdscy0CSdAbwtIurWZg0kfQKY0AmfE0n7khL+VrWa2MpidamR2hpG0g6Sds1NC3uRrry4qt1xlZ2kwcC/kC7zK7XcdPBZ4LwyJ1FwIrXOtT7psHMp6Uz0mcCv2hpRyeV23SdI7bGX9LJ6W0nakXSuYFNS012ryp0i6W+S7uthuSSdLWmepHsk7dFQuT60N7M1RW4qeB74cUS8oR1eqR+AE0k3VuwNfC8iqtuY38A1UjNbY0TEDNI10z0ZR0qyERG3ka5H7+lE8Wva3WFHW2ntrmDdzvgTjNhqy95XKpG1unq6F6Kc1u5et90hNEw9Xh5dPo88/AhLFi9pKmANWzdY1mAT6XOvzCFdjlUxOSL60h68GSvf3LEwz3us3os6I4v0l3XXgr1r3nBUOj+44DvtDqFP3rzOm9sdQp+8Y4MR7Q6hYWtpULtDaNh+Yw5ovpBlrzb+Pf39oy9FxOjeV+xRraTfa/vnmp1IzawzaMBq4QtZ+S654aRruOtyG6mZlZuAbjX2aN5U4BP57P27gGfy3Yp1uUZqZuXXogqppEtJPXYNk7SQdDv5IICImES6LfZQ0p1sL5C6WuyVE6mZlZxadmgfEXU7xMl9HPxrX8t1IjWzchOlb4R0IjWz8hu4k02rxInUzMqv3HnUidTMSq5y1r7EnEjNrPx8aG9m1qRy51EnUjMrOQFd5c6kTqRmVn7lzqNOpGZWchJ0l/tCUidSMys/10jNzJrks/ZmZk0qdx51IjWzkuuAs/b91oIr6buSTipMXyvpvML0mZI+J+maFm1vYh75LyQNa0WZZlYSavDRJv15KuwWYCyApC5gGLBTYflYcj+ALXIz8F7g4RaWaWZlMHAdO6+S/kykN5MTKSmB3gc8J2kjSesAOwJ3AUMk/VzSA5IullKrsqQ9Jd0oaVauzW4qaRtJd1Y2IGmEpFkAEXFXRPylH/fHzNpBavzRJv3WRhoRiyQtl7QFKaHeShqNbwzwDHAPsAzYnZRoF5GS77sl/RH4PjAuIp6QdCTwtYg4TtIzkkZFxGxS79UX9iUuSROACQCs21kjXZqtscrdRNrv3aVWaqWVRHprYfqWvM7tEbEwIl4FZgNbAdsDOwO/kzQbOI00CBXAecB4Sd3AkcAlfQkoIiZHxOiIGM2gcl/ka2ZZC2ukkg6WNDefUzmlxvKNJF0l6R5Jt0vaubcy+/usfaWddBfSof0C4GTgWWBKXuflwvorckwC5kTEmBpl/oI0zsr1wKyIWNI/oZtZabSozpMrYOcAB5FGDJ0paWpE3F9Y7T+A2RFxhKQd8voHDkB4PboZOAx4MiJWRMSTwFDS4f2tdV43F9hY0hgASYMk7QQQES8B1wI/BC7ox9jNrAwqlz818ujdXsC8iJgfEcuAy4BxVeuMBP4AEBEPAFtJ2qReof2dSO8lna2/rWreMxGxuKcX5R38MHCGpLtJh/xjC6tcDARwXWWGpM/kUQGHA/cUL7Uysw7XeCIdJumOwmNCVUmbkY6MKxbmeUV3Ax8CkLQXsCWvNy3W1K+H9hGxAtigat6xhefTgemF6YmF57OBfXsoeh9gSi6/sv7ZwNnNR21mpdP4GfnFETG6Xkk15kXV9DeA7+XzM/eSri5aXm+jHXdnk6SrgG2AA9odi5kNgNZebL8Q2LwwPZx0xdBrIuJZ8nj2+XLMh/KjRx2XSCPiiHbHYGYDSajBGml11bKGmcAISVsDjwJHAR9daWvSUOCF3MR4PDAjJ9cedVwiNbM1T6sSaUQslzSRdMK6m9REOEfSCXn5JNLNQj+WtAK4H/hkb9t1IjWzUhPQ3WCnJa82sE5ETAOmVc2bVHh+KzCiDyE6kZpZyanxGmm7OJGaWek5kZqZNaXxk03t4kRqZqVX8jzqRGpm5SZ8aG9m1hxBl8rdU5sTqZmVnmukZmZNKnkedSI1s3IToqvkmdSJ1MxKz4f2ZmbNEHSVfFx7J1IzKzVf/lRyO22zLT/7WZ/Gzmubnf7psHaH0CePXHVju0Pok251zldh7e512h1Cw1rVtulEambWFN8iambWHPf+ZGbWvJLnUSdSMys3AV1d5b5FtNzRmZmRTlo18miEpIMlzZU0T9IpNZZvKOlqSXdLmiNpfK/xrcI+mZkNHKVD+0YevRYldQPnAIcAI4GPSBpZtdq/AvdHxG7A/sCZktauV64TqZmVmvJZ+0YeDdgLmBcR8/MooZcB46rWCWD9PBTzEOBJVrdx7c1szaPGB7YfJumOwvTkiJhcmN4MWFCYXgjsXVXGD4CppPHu1weOjIi64+o5kZpZ6fXh8qfFETG6XlE15lWP4vx+YDZwALAN8DtJN9Ub296H9mZWel1daujRgIXA5oXp4aSaZ9F44MpI5gEPATvUja8P+2JmNuCUL8hvURvpTGCEpK3zCaSjSIfxRY8AB6ZtaxNge2B+vUJ9aG9mJde6W0QjYrmkicC1QDcwJSLmSDohL58EfAW4UNK9pKaAL0TE4nrlOpGaWem18hbRiJgGTKuaN6nwfBHwvr6U6URqZqXnW0TNzJoglf8WUSdSMys99/5kZtakkufR/rv8SdJ3JZ1UmL5W0nmF6TMlfU7SNS3a3sW5I4L7JE2RNKgV5ZpZu7X0FtF+0Z8ND7cAYwEkdQHDgJ0Ky8cCrUx2F5Mumt0FWA84voVlm1kbrcmJ9GZyIiUl0PuA5yRtJGkdYEfgLmCIpJ9LeiDXKgUgaU9JN0qalWuzm0raRtKdlQ1IGiFpFqRLGvKdCAHcTrpjwcw6XIsvyO8X/dZGGhGLJC2XtAUpod5K6jBgDPAMcA+wDNidlGgXkZLvuyX9Efg+MC4inpB0JPC1iDhO0jOSRkXEbNKtXBcWt5sP6Y8GPttf+2ZmA2tNH465UisdC3yHlEjHkhLpLXmd2yNiIYCk2cBWwNPAzqTOAiDdgfBYXv88YLykzwFHkrrFKjoXmBERN9UKSNIEYALApsM3bXL3zGxAlPxsU38n0ko76S6kQ/sFwMnAs8CUvM7LhfVX5JgEzImIMTXK/AXwZeB6YFZELKkskPRlYGPgUz0FlLvUmgyw8+4jq3t9MbPSKf8oov19levNwGHAkxGxIiKeBIaSDu9vrfO6ucDGksZAOlyXtBNARLxEuk/2h8AFlRdIOp7U/dVHeus70Mw6SAt7yO8v/Z1I7yWdrb+tat4z9ToByD1Xfxg4Q9LdpL4BxxZWuZjUh+B1hXmTgE2AWyXNlvSlluyBmbWVWINPNgFExApgg6p5xxaeTwemF6YnFp7PBvbtoeh9SL22rCis75sLzFZTZT+077jkI+kqUq/VB7Q7FjMbGGv6WfuWi4gj2h2DmQ2gNh+2N6LjEqmZrVkqbaRl5kRqZqVX9kRa7k7+zMxo7Vl7SQfnDo7mSTqlxvLP5yt/ZudOkFZIenO9Mp1Izazc1LpRRCV1A+cAhwAjgY9IGllcJyK+FRGjImIU8EXgxnwNfI+cSM2s1NTabvT2AuZFxPx8vfplwLg6638EuLS3Qp1Izaz0+pBIh0m6o/CYUFXUZqRb1SsW5nm1tjkYOJh0W3pdPtlkZqXXh3NNiyNidL2iaszrqc+Nw4GbezusBydSMys7tfSs/UJg88L0cFIXnrUcRQOH9eBDezPrBK3rtWQmMELS1pLWJiXLqW/cnDYE9gN+1UihrpGaWakJ6G7RLaIRsVzSRFIPct2kPjvmSDohL5+UVz0CuC4iljZSrhOpmZVca28RjYhpwLSqeZOqpi+kavSNepxIzazcBF0lv7PJidTMSs332puZtUDZz4qv0Yl0+asreHrZU+0OoyG3/PSn7Q6hT7Y4Yr92h9AnS665s/eVSmLZipd7X6kkXo3mh0VLJ5vKnUrX6ERqZp1AbiM1M2tKay/I7xdOpGZWasJtpGZmTfOhvZlZk3xob2bWBAHdq0sizT1Lb1J8TUQ80h9BmZm9bjU5ay/pRODLwOPAq3l2ALv2U1xmZkDq1Gm1SKTAZ4HtI2JJfwZjZlbL6tJGugB4pj8DMTPryepSI50PTJf0a+C1+9Mi4jv9EpWZWSZqjw9SJo0m0kfyY+38MDMbIGKt1eFe+4j4r/4OxMysFnX6LaKSzoqIkyRdTY2R9iLi7/stMjOzrJVtpJIOBr5HGmrkvIj4Ro119gfOAgaRRiat251ZbzXSn+T/v93HWM3MWqZVaTRfD38OcBBpRNGZkqZGxP2FdYYC5wIHR8Qjkt7aW7l1E2lEzMr/3yhpPWCLiJi76rthZtY3oqU10r2AeRExH0DSZcA44P7COh8FrqzccBQRf+ut0IZacCUdDswGfpunR0l6wxCmZmatJ7q7uhp6AMMk3VF4TKgqbDPS5ZwVC/O8ou2AjSRNlzRL0id6i7DRs/ankzL5dICImC1pqwZfa2a2yvrYjd7iiBjdS3HVqs//rAXsCRwIrAfcKum2iHiwp0IbTaTLI+KZvpw5k/Rd4OGIOCtPXwssiIjj8/SZwKPAARFxWMMF97y984HRpD/Ug8CxEfF8s+WaWZu19qz9QmDzwvRwYFGNdRbnMe2XSpoB7EbKKzU1mujvk/RRoFvSCEnfB27p5TW3AGMBJHUBw4CdCsvHks6Itcq/RcRuEbEr6ZrXiS0s28zaqEtq6NGAmcAISVtLWhs4CqhupvwV8HeS1pI0GNgb+FPd+BrcjxNJSfBl4BLS7aKf7eU1N5MTaX7tfcBzkjaStA6wI3AXMETSzyU9IOli5Z8eSXtKujG3UVwraVNJ20h6bZSynNQrJ8SezfNEqo43P+qWmbVd5WRTKxJpRCwnVbKuJSXHn0XEHEknSDohr/Mn0vmge4DbSZdI3Vev3EYP7T8QEacCp762c9I/AlfUCXiRpOWStiAl1FtJjbpjSIn4HmAZsDsp0S4iJd93S/oj8H1gXEQ8IelI4GsRcZykZySNiojZwHjgwkJMFwCHks7AnVwrrtz4PAFgk+GbNLj7ZtZOrbwgPyKmAdOq5k2qmv4W8K1Gy2y0RvrFBudVq9RKK4n01sJ0pWng9ohYGBGvkq4M2ArYHtgZ+J2k2cBppLYMgPOA8fl6sCNJNWQAImI88HbSL82RtQKKiMkRMToiRg9984YN7IKZtZfoVldDj3bp7c6mQ0g1vM0knV1YtAGwvIHyK+2ku5AO7ReQaorPAlPyOsVBulfkmATMiYgxNcr8Balv1OuBWdVd+0XECkmXA58HLmggRjMrsU7oj7S3FL4IuAN4CZhVeEwF3t9A+TcDhwFPRsSKiHgSGEo6vL+1zuvmAhtLGgMgaZCknQAi4iVS+8YPyYlSybaV58DhwAMNxGdmHUAN/muX3u5suhu4W9IlpFridnnR3Ih4pYHy7yWdrb+kat6QiFjcU7tHRCyT9GHgbEkb5jjPAubkVS4GPgRcl6cFXCRpg/z8buDTDcRnZh2gozstKRgL/Bj4CylRbS7pmIiYUe9FEbGC1AxQnHds4fl08kX+eXpi4flsYN8eit4HmJLLJ7evvrvBfTGzDqLVZcwm4DvA+yr32UvaDriUdPX/gJJ0FbANcMBAb9vM2kN9ubepDRpNpIOKnZVExIOSWnkxfcMi4oh2bNfM2qd7dejYGZiVb8GsdKv3MdJJJzOzftXuE0mNaDSRngD8K/AZUhvpDFJ/fWZm/asDLn/qNZHm++RnRcTOpLZSM7MBVfaz9r02POQz4nfnWz3NzAZU6kavsX/t0uih/abAHEm3A0srMz1mk5n1P9G1mpxs8iiiZtY2XZ18sknSuqQTTduS7kg6P3dDZWY2IET520h7q5FeBLwC3AQcAoyk935IzcxaZzU4az8yInaB14byuL3/QzIzK+r860hf65gkIpaXvXptZquf1EN+uU829RbdbpKezY/ngF0rzyU9OxABmpl1qauhRyMkHSxprqR5kk6psXz/PBLH7Pz4Um9l9taNXndDkZmZ9ZvW9f6UR9Y4BziINFroTElTI+L+qlVv6svoxuWuL5vZGk+0tGPnvYB5ETE/IpYBlwHjmo3RidTMSq8Po4gOk3RH4TGhqqjNSEMeVSzM86qNkXS3pN9URueop9EL8ldLz7+ylJseva3dYTRk7pJF7Q6hT87+Ts1BXEvrLYft0e4QGjb7sivbHULDlq14qflCBGr8ZNPiiBhdv7Q3qB66/U5gy4h4XtKhwC+BEfU26hqpmZVcowf2DR3aLwQ2L0wPJ41N95qIeDYins/PpwGDJA2rV+gaXSM1s/ITLe3YeSYwQtLWwKPAUcBHV9qe9Dbg8YgISXuRKpxL3lBSgROpmZVeq+61z9fDTySNRNxNGvttjqQT8vJJwIeBT0taDrwIHBUR1Yf/K3EiNbNSa/W99vlwfVrVvEmF5z8AftCXMp1Izazk1JeTTW3hRGpmpdfR3eiZmbWbVP577Z1Izazk1PH9kZqZtZ0P7c3MmpDO2vvQ3sysCZ3fsbOZWdu5jdTMrEk+a29m1gThk01mZs2RL38yM2uaSt7jpxOpmZVe2WukpUjzkk6VNEfSPXnUvr1XoYyvSVog6fn+iNHM2kOIbnU19GiXttdIJY0BDgP2iIiXc0/Ua69CUVeTur76cyvjM7P2K/t1pGWokW5KGmflZYCIWBwRiyQdKOkuSfdKmiJpnTzvqsoLJR0k6cr8utsi4rE27YOZ9SPlE069PdqlDIn0OmBzSQ9KOlfSfpLWBS4EjoyIXUg1508D1wM7Sto4v3Y8cEE7gjazgZGGY+5q6NEubU+keZCpPYEJwBPA5cCngIci4sG82kXAvrm7/58AH5c0FBgD/KYv25M0oTJU69KnlrZoL8ys/zRWG220RirpYElzJc2TdEqd9d4paYWkD/dWZtvbSAEiYgUwHZgu6V7gmDqrX0BqD30JuCIilvdxW5OByQDDRw6vOw6LmZVDqy7Il9QNnAMcRBpRdKakqRFxf431ziCN7dRAfG0maXtJxTGjRwGPA1tJ2jbPOxq4ESAiFpGGTz2NdPhvZquxSsfOjTwasBcwLyLmR8Qy4DJgXI31TgR+AfytkULbnkiBIcBFku6XdA8wEjiF1P55Ra6hvgpMKrzmYmBB8VdE0jclLQQGS1oo6fQB2wMz61d9OLQfVmm6y48JVUVtBiwoTC/M84rb2gw4gpVzTl1tP7SPiFnA2BqL/gDs3sPL9gF+VFXOvwP/3trozKz91JcTSYsjYnTdwt6ouonvLOALEbGi0XbXtifSvpI0C1gKnNzuWMxsYHS17tKmhcDmhenhpKbCotHAZZUaLnCopOUR8cueCu24RBoRe7Y7BjMbOOnyp5Yl0pnACElbA48CRwEfLa4QEVu/tm3pQuCaekkUOjCRmtmap1UX20fEckkTSWfju4EpETFH0gl5ecPtokVOpGZWcmppx84RMQ2YVjWvZgKNiGMbKdOJ1MxKLXXsXIYLjHrmRGpm5abyd6PnRGpmJedRRM3MmuYaqZlZE9xGambWCq6Rmpk1w22kZmZNcxupmVmTXCM1M2uSE6mZWRPU4ltE+4MTqZmVnmukJfZqvMrSV15sdxgNWX+dddodQp8sXdYZf9eKP1z4o95XKolRR32o3SE0bn5DI3XU51tEzcya5xqpmVkThGukZmZN8gX5ZmZNK/tZ+3JHZ2ZGpU7a+7+GypIOljRX0jxJp9RYPk7SPZJm5yGd9+mtTNdIzazUWjn4naRu4BzgINKIojMlTY2I+wur/QGYGhEhaVfgZ8AO9cp1jdTMSk5IjT0asBcwLyLmR8Qy4DJgXHGFiHg+Iipj3b+JN457/wZOpGbWAdTgg2H5cLzymFBV0GbAgsL0wjxv5a1JR0h6APg1cFxv0fnQ3szKTX062bQ4IkbXL+0N3lDjjIirgKsk7Qt8BXhvvY26RmpmpdfCk00Lgc0L08OBRT2tHBEzgG0kDatXqBOpmZWaWttGOhMYIWlrSWsDRwFTV9qetK1yYZL2ANYGltQr1If2ZlZ6rTprHxHLJU0ErgW6gSkRMUfSCXn5JOAfgE9IegV4ETiycPKpJidSMyu9Vt7ZFBHTgGlV8yYVnp8BnNGXMp1Izaz0fK+9mVkT3LGzmVkLuNMSM7OmOZGamTWl3Gm0JNeRSjpV0pxCjyt79/H1gyX9WtIDuZxv9FesZjbwWngdab9oe41U0hjgMGCPiHg530Gw9ioU9e2IuCFfZPsHSYdExG9aGqyZtUm566RlqJFuSro/9mWAiFgcEYskHSjpLkn3SpoiaZ0876rKCyUdJOnKiHghIm7Ir18G3Em69cvMOl6jN4i2L9mWIZFeB2wu6UFJ50raT9K6wIWkOwp2IdWcPw1cD+woaeP82vHABcXCJA0FDif1KfgGkiZUeoZ54ekX+mWHzKx1pPIf2rc9kUbE88CewATgCeBy4FPAQxHxYF7tImDffJvWT4CP54Q5Bnjt8F3SWsClwNkRMb+H7U2OiNERMXrw0MH9tFdmtiZpexspQESsAKYD0yXdCxxTZ/ULgKuBl4ArImJ5Ydlk4M8RcVY/hWpmbVD260jbXiOVtL2kEYVZo4DHga0kbZvnHQ3cCBARi0jdXp1GOvyvlPNVYEPgpH4P2swGVNnbSMtQIx0CfD8fqi8H5pEO8y8FrsiH6zOBSYXXXAxsXBlnRdJw4FTgAeDO3Fbyg4g4b6B2wsz6j++170VEzALG1lj0B2D3Hl62D/CjQhkLKfv1EWa22mp7Iu0rSbOApcDJ7Y7FzAZCew/bG9FxiTQi9mx3DGY20JxIzcxW2Wvjg5ZY28/am5n1ppUX5Es6WNJcSfMknVJj+cdyvx/3SLpF0m69lekaqZmVXqvaSCV1A+cAB5FGFJ0paWrlCqDsIWC/iHhK0iGk69PrdqTkGqmZdQA1+OjVXsC8iJif++W4DBhXXCEibomIp/LkbTTQb4drpGZWcn26j36YpDsK05MjYnJhejNgQWF6IfVrm5+kcBt6T5xIzWx1sjgiRtdZXisj1xxqWdJ7SIl0n9426kRqZqWWDtpbdt5+IbB5YXo46Zbzlbcp7QqcBxwSEUt6K9RtpGbWAVrWRjoTGCFp69wJ/FHA1JW2JG0BXAkcXeiBri7XSM2s9LpadK99RCyXNBG4FugGpkTEHEkn5OWTgC8BbwHOzW2zy3tpLnAiNbOya+0l+RExDZhWNW9S4fnxwPF9KdOJ1MxKr+x3NjmRmlkHKHcqdSI1s3KT+yM1M2tKiy9/6hdK48mtmSQ9ATzcD0UPAxb3Q7n9oZNihc6Kt5Nihf6Jd8uI2Lj31Xom6bek2BqxOCIObmZ7q2KNTqT9RdIdvV0uURadFCt0VrydFCt0Xrxl4gvyzcya5ERqZtYkJ9L+Mbn3VUqjk2KFzoq3k2KFzou3NNxGambWJNdIzcya5ERqZtYkJ9Iqkr4r6aTC9LWSzitMnynpc5KuadH2JuZBuEJSo9fKtSvWi/OgYfdJmiJpUMnjPV/S3XkQs59LGtKCMk+VNCeXOVtS3bF8eijja5IWSHq+2Xga2FZT8UoaLOnXkh7I5Xyjv2LtZE6kb3QLMBZAUhfpQuCdCsvHAn1KIL24GXgvq3ZjwEDHejGwA7ALsB597CGHgY/33yJit4jYFXgEmNhMYZLGAIcBe+Qy38vKw1Y06mrS2EH9qoXxfjsidgB2B96dB4SzAifSN7qZ/GUnfcnvA56TtJGkdYAdgbuAIbmW80CuqQlA0p6SbpQ0K9e4NpW0jaQ7KxuQNELSLICIuCsi/tIhsU6LDLidBgYFa3O8z+Z5IiX+Zs+sbkq6c+blXP7iiFgk6UBJd0m6N9fU18nzrirEdZCkK/PrbouIx5qMZUDijYgXIuKG/PplwJ30/X1f/UWEH1UP4C/AFsCngBOArwCHAu8GZgD7A8+QPlBdwK2kcV0GkWpdG+dyjiR1HAtwAzAqP/86cGKNbQ7rkFgHkb5Qf1f2eIELgMfzOoOb/FwMAWYDDwLnAvsB65JqedvldX4MnES6RfyBQryXAIdXlfd8P3+OWx3vUGA+8I52f0fL9nCNtLZKzWks6Yt8a2H6lrzO7RGxMCJeJX1YtwK2B3YGfidpNnAar/96nweMVxpX+0jSB7VTYz0XmBERN5U93ogYD7wd+FNetsoi4nlgT2AC8ARwOekH4aF4fUiKi4B9I2WenwAflzQUGEMDo1G2UivjlbQWcClwdkTMH7Cd6BDu/am2SlveLqTDzwXAycCzwJS8zsuF9VeQ/pYC5kTEmBpl/gL4MnA9MCsaGFCrjLFK+jKwMekLWfp4ASJihaTLgc+TaqirLCJWANOB6ZLuBY6ps/oFpPbQl4ArImJ5M9teFS2MdzLw54g4q59C7WiukdZ2M6mR/smIWBERT5IOa8aQalA9mQtsnBv5kTRI0k4AEfESaZyYH9Lkl7ldsUo6Hng/8JFcWyxtvEq2rTwHDicduq4ySdtLGlGYNYrUbLBVZVvA0cCNOa5FpBEqTwMubGbbq6JV8Ur6KrAhqQnAanAire1e0hnl26rmPRMRPXYzFqkx/sPAGZLuJh2Wji2scjHphMd1lRmSPiNpIekw9R4VLgcqW6zAJGAT4FalS2m+1MdYBzJeARflWti9pBMv/70K8RYNyWXeL+keYCRwCjAeuCJv61XS36kY14KIuL8yQ9I383s+WNJCSac3GVe/xStpOHBqfu2d+X3v69Uaqz3fIjqAJP0fYMOI+M92x9KbTooVyhuvpB8Ad0XE+e2OpRGdFm9ZOJEOkHxpyTbAAfVqXmXQSbFCeePNl2EtBQ6KfAlSmXVavGXiRGpm1iS3kZqZNcmJ1MysSU6kZmZNciK1ASNpRb585j5JV0ga3IfXjpJ0aAPrjZZ0dnORmvWNE6kNpBcjYlRE7AwsI91r/5p8i2dPRpHuya8rIu6IiM80FaVZHzmRWrvcBGwraX9JN0i6BLhX0rqSLsg9E90l6T2S1iZdTH9krtEeKelNueeimXm9cQC5vGvy89PzOtMlzZfkBGv9wvfa24DLHWAcAvw2z9oL2DkiHpJ0MkBE7CJpB9KdStsBXwJGR8TEXMbXgesj4rjcycbtkn5fY3M7AO8B1gfmSvphRLzSj7tnayDXSG0grZd7brqD1NFy5e6Z2yPiofx8H1IvRETEA6QOr7erUdb7gFNyedNJ3cNtUWO9X0fEy/lC/b+RbnE1aynXSG0gvRgRo4ozUn8iLC3OarAsAf8QEXOryqtOlLV6kjJrKddIrWxmAB8DkLQdqZY5F3iOdHhecS1wYu7ZCUm7D3CcZq9xIrWyORfozj0TXQ4cm+/7vgEYWTnZROpZfxCpx6z78rRZW/heezOzJrlGambWJCdSM7MmOZGamTXJidTMrElOpGZmTXIiNTNrkhOpmVmT/j+JT6bQS4+LfAAAAABJRU5ErkJggg==\n",
      "text/plain": [
       "<Figure size 432x288 with 2 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#Import matplotlib.pyplot to help with setting up the heatplot\n",
    "import matplotlib.pyplot as plt\n",
    "\n",
    "#Looks at the numpy array with pairwise sequnence alignment scores will be helpful with data analysis\n",
    "print(heatmapscore)\n",
    "#Normalize score for heatmap\n",
    "#Source: https://stackoverflow.com/questions/29301904/extracting-minimum-values-per-row-using-numpy\n",
    "#Source: https://stackoverflow.com/questions/19602187/numpy-divide-each-row-by-a-vector-element\n",
    "#Creates an array \"vector\" with maximum value of \"heatmapscore\" for each column\n",
    "vector = heatmapscore.max(axis=0)\n",
    "#Code check to see that it gives the maximum value \n",
    "#print(vector)\n",
    "\n",
    "#Divides each column in \"heatmapscore\" with corresponding value in \"vector\" or the maximum value for each column\n",
    "normalize = heatmapscore / vector[None,:]\n",
    "#Code check\n",
    "#print(normalize)\n",
    "\n",
    "#Creates subplot \n",
    "fig, ax = plt.subplots()\n",
    "\n",
    "#Actually show the heatplot\n",
    "plt.imshow(normalize)\n",
    "\n",
    "#Shows the axes as well as able to manipulate the color of the heatmap\n",
    "im = ax.imshow(normalize, cmap = 'Greens')\n",
    "\n",
    "#Adds a gradient scale on the right where gradient of colors have corresponding values associated aka the score\n",
    "cbar = ax.figure.colorbar(im)\n",
    "\n",
    "#Adds a title to the heatmap\n",
    "plt.title(\"Pairwise Sequence Alignment Whey vs Soy\")\n",
    "\n",
    "#Adds a title to the x-axis\n",
    "plt.xlabel(\"Protein\")\n",
    "\n",
    "#Adds a title to the y-axis\n",
    "plt.ylabel(\"Protein\")\n",
    "\n",
    "#Aim is to label the xticks and yticks on the heatmap\n",
    "#Source: https://stackoverflow.com/questions/3100985/plot-with-custom-text-for-x-axis-points\n",
    "#Global variable \"x\" which is an array of the original x-ticks\n",
    "x = np.array([0,1,2,3,4])\n",
    "#Global variable \"y\" which is an array of the original y-ticks\n",
    "y = np.array([0,1,2,3,4])\n",
    "#GLobal variable \"my_xticks\" which is an array that will replace the original x-ticks\n",
    "my_xticks = ['Whey1','Whey2','Whey3','Soy1','Soy2']\n",
    "#GLobal variable \"my_yticks\" which is an array that will replace the original y-ticks\n",
    "my_yticks = ['Whey1','Whey2','Whey3','Soy1','Soy2']\n",
    "#Replaces original x-ticks \"x\" with \"my_xticks\"\n",
    "plt.xticks(x, my_xticks)\n",
    "#Replaces original y-ticks \"y\" with \"my_yticks\"\n",
    "plt.yticks(y, my_yticks)\n",
    "\n",
    "#Beta-lactoglobulin = Whey1\n",
    "#Alpha-lactalbumin = Whey2\n",
    "#Glycomacropeptide = Whey3\n",
    "#β-conglycinin = Soy1\n",
    "#Glycinin = Soy2 \n",
    "\n",
    "#Show the plot\n",
    "plt.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "03f744ec",
   "metadata": {},
   "source": [
    "### Analysis of the Results\n",
    "The heat plot above shows the pairwise sequence alignment scores for each protein pair as labelled. The data was normalized for each column and so we will be considering the majority of the data row by row. The data shows proteins that are more similar to one another have a higher normalized scores whereas proteins that are not as similar will have lower normalized scores. The first row looks at the relationship of Whey1 with all other proteins and we can see that there is a decrease in the normalized score as we go down from Whey2 to Soy2. Roughly the normalize similarity score between Whey1 and Whey2 is between 0.4 and 0.5, between Whey1 and Whey3 it is between 0.3 and 0.4, between Whey1 and Soy1 it is between 0.2 and 0.3, and between Whey1 and Soy2 it falls below 0.2. However, interestingly when you look at the first column you see almost the opposite and this is attributed to the fact that if we take the raw pairwise scores Soy1 and Soy2 show higher scores compared to Whey2 and Whey3 simply due to the fact that their amino acid sequence is signficantly longer but when we normalize the data we get something that is closer to what the first row depicts. The same relationship is shown within the second row and second column and the third row and third column where the second and third row show a decrease in the normalized scores as we go from left to right. Meanwhile the second and third column show a increase in the normalized scores as we go from top to bottom because of what we mentioned before regarding how the raw pairwise scores are higher because of Soy1 and Soy2's sheer length. It should be noted that there are slight differences in the values for the normalized scores when we look at the second row and third row. Interestingly there is a shift when we get to the fourth row and fifth row compared to the fourth column and fifth column. The fourth and fifth row again show the problem where there is an indication of a higher normalized score but the lengths of the longer proteins have not been taken into full consideration. Whereas, the fourth and fifth column show more accurate representations of the normalized scores between each protein. Overall, the whey proteins show an increase in their normalized scores when being compared to one another and a decreased in their normalized scores when being compared to soy proteins. Likewise, soy proteins show an increase in their normalized scores when being compared to one another and a decrease in their normalized scores when being compared to whey proteins. However, the question remains whether the normalized scores are sufficient to determine if there are little to no overlaps between whey and soy proteins as there is evidence of some shared similarity but just tends to range from slightly below 0.2 to 0.3. "
   ]
  },
  {
   "cell_type": "markdown",
   "id": "a15584b7",
   "metadata": {},
   "source": [
    "### References (Note: Can see where each link was used by looking at comments in code or citations in other sections)\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5124466/\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5934683/\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6471465/\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6164536/\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/pmc/articles/PMC2682512/\n",
    "\n",
    "https://www.w3schools.com/python/numpy/numpy_random.asp\n",
    "\n",
    "https://www.activestate.com/resources/quick-reads/what-is-matplotlib-in-python-how-to-use-it-for-plotting/\n",
    "\n",
    "https://biopython.org/wiki/AlignIO\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/CAY39357.1 \n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/BAC06860.1\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/P02668\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/AAB23463.1\n",
    "\n",
    "https://www.ncbi.nlm.nih.gov/protein/AAA33964.1\n",
    "\n",
    "https://www.biostars.org/p/18129/\n",
    "\n",
    "https://stackoverflow.com/questions/31152011/multiple-sequence-alignment-with-unequal-string-length\n",
    "\n",
    "https://www.biostars.org/p/142519/\n",
    "\n",
    "https://towardsdatascience.com/pairwise-sequence-alignment-using-biopython-d1a9d0ba861f\n",
    "\n",
    "https://canvas.ucsd.edu/courses/29418/files/5721198?module_item_id=1126617&fd_cookie_set=1\n",
    "\n",
    "https://stackoverflow.com/questions/29301904/extracting-minimum-values-per-row-using-numpy\n",
    "\n",
    "https://stackoverflow.com/questions/19602187/numpy-divide-each-row-by-a-vector-element\n",
    "\n",
    "https://stackoverflow.com/questions/3100985/plot-with-custom-text-for-x-axis-points"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
